### ISSUE TITLE
Document copy/paste shortcuts in terminal

### ISSUE THREAD

**Issue by timothyclemansinsea** [jupyterlab/jupyterlab#1146]
Triage summary: Document the various keyboard shortcut idiosyncracies for cutting/pasting into the terminal.

Or even better? Implement William's method noted at https://github.com/jupyterlab/jupyterlab/issues/1146#issuecomment-426436903

---
Original title: Make copy/paste work in terminal

I just started using Jupyter Lab. In SageMathCloud terminals one can copy paste with ctrl-c/ctrl-v. Ctrl-c is copy if text is selected. I just tried to copy and paste and couldn't at all.

**Comment by blink1073** [comment]
Hi @timothyclemansinsea, can you describe your setup?  Using the latest development branch on OSX 10.11 in Firefox and Chrome I am able to do what you are describing.

**Comment by timothyclemansinsea** [comment]
@blink1073 I'm using Chrome on Windows. Seems I can copy using right click but not paste.

**Comment by blink1073** [comment]
Is your notebook server running on a POSIX machine, or do you have terminals working natively on Windows?

**Comment by buddiex** [comment]
I still can paste on the terminal.
my note book is serving from a Linux server.
Right clicking gives only the option "Refresh Terminal" no other options to paste or copy.
and keyboard shortcuts don't work.

**Comment by blink1073** [comment]
Hi @buddiex, you can get the system right click menu by using shift+right click.

**Comment by apcamargo** [comment]
I can't copy/paste with the keyboard shortcuts either

**Comment by FinLouarn** [comment]
I can't copy paste with ctrl+c/ctrl+v (Ubuntu 16.04 on local laptop, Firefox 55.0.2)

**Comment by blink1073** [comment]
Ah, I think the issue is that Ctrl+C is typically used in Unix to interrupt a process in the terminal.

**Comment by keheyndr** [comment]
True, but what normally work is Shift+Insert to paste. And as a matter of fact, that does work. 

What is missing for a full (quick) terminal experience is that selection text using your mouse automatically pastes text to the clipboard.

**Comment by qingpeng** [comment]
shift + insert works for windows, but what if I am . using Mac? where is the Insert key?

**Comment by allenyllee** [comment]
My Firefox 59.0.2 doesn't have copy/paste items in context menu when triggering `shift + right click`. 
But keyboard shortcut `Shift + Insert` works.
Also, `Ctrl+Insert` copy selected text to clipboard.

**Comment by danijel3** [comment]
Typically copying pasting into and from a GUI terminal in Linux is done using "shift-ctr-c" and "shift-ctrl-v". I think that would be a great addition to the jupyter terminal as well. Is it possible to configure keyboard shortcuts on your own?

**Comment by Honghe** [comment]
JupyterLab terminal copy/paste shortcut key based on the server side we connect. Such as:
* server side is Windows 10, then the terminal connects to PowerShell, copy/paste is `ctrl+c`/`ctrl+v`.
* server side is Ubuntu, then the terminal connects to our set bash/zsh, copy/paste is `ctrl+shift+c`/`ctrl+shift+v`.

**Comment by jasongrout** [comment]
I suggest we verify the note above (https://github.com/jupyterlab/jupyterlab/issues/1146#issuecomment-394686488) and add it to the documentation to explain the shortcuts, and then close this issue.

**Comment by nfultz** [comment]
That seems a little unusual, I would have expected it to match the _client_ OS shortcuts, eg whatever the web browser is running on, not the remote server.

**Comment by jasongrout** [comment]
@williamstein investigated this behavior of xterm in depth and proposed a solution he uses in cocalc with term.js: https://groups.google.com/d/msg/jupyter/3t1WEY_FFEg/b-NbW3rNCAAJ

**Comment by jasongrout** [comment]
VSCode documents the different shortcuts: https://code.visualstudio.com/docs/editor/integrated-terminal#_copy-paste

**Comment by tgeorgeux** [comment]
The way I see it we have two issues:
1) An education issue about how to access the browser context menu.
2) A conflict with Linux machines.

@jasongrout One thing I didn't pick up from William's comments, was how feasible will it be to accomplish what he's done in term.js? I think his solution is quite good. 

If it's not feasible to integrate what he's proposed, we can add both accel-c/v and accel-shift-c/v as defaults for copy in the terminal. 

Either of these solutions address the second issue but not the first.

**Comment by williamstein** [comment]
> @jasongrout One thing I didn't pick up from William's comments, was how feasible will it be to accomplish what he's done in term.js? I think his solution is quite good.

I didn't know how feasible it was until I did it.   It turns out this is one place where xterm.js's public API works really well.  Here's basically all you have to do:
```
    terminal.attachCustomKeyEventHandler(event => {
      if (
        (event.ctrlKey || event.metaKey) &&
        event.key === "c" &&
        terminal.hasSelection()
      ) {
        // Return so that the usual OS copy happens
        // instead of interrupt signal.
        return false;
      }

      if ((event.ctrlKey || event.metaKey) && event.key === "v") {
        // Return so that the usual paste happens.
        return false;
      }

      return true;
    });
```

I looked up [who has been doing the most work on xterm.js recently](https://github.com/xtermjs/xterm.js/graphs/contributors), and it's https://github.com/Tyriar (@Tyriar); he just happens to be at some local company near me called Microsoft, so I'm meeting with him there tomorrow (Wed, Oct 3)...

**Comment by williamstein** [comment]
Regarding

> 1. An education issue about how to access the browser context menu.

There could possibly be some subtlety with that.   The problem is that at least in some configurations of xterm.js on some browsers, the context menu doesn't appear at all. For example, xterm.js integration is now live in https://cocalc.com, and when I use it via my Chromebook (using ChromeOS stable channel), I do NOT see the context menu.  See screenshot below.  This may have to do with using the canvas renderer instead of the DOM renderer.   I ended up going with the canvas render, since measuring the size is very robust, and scrollback works perfectly, and copy selection also works perfectly (these are all pretty broken with the DOM renderer for me).

![screenshot 2018-10-02 at 4 11 59 pm](https://user-images.githubusercontent.com/1276278/46382093-2d0f4100-c65e-11e8-92ce-024c5441260d.png)

In any case, when documenting/education about the context menu, be sure to point out "this might not work for you", if that's the case.

**Comment by Tyriar** [comment]
Yes, [`Terminal.attachCustomKeyEventHandler`](https://github.com/xtermjs/xterm.js/blob/d4ec50e7de46e958ee499d629bcf131d59a694a5/typings/xterm.d.ts#L461-L470) was added to allow people to intercept any key they want before it's evaluated by xterm.js.

**Comment by fonnesbeck** [comment]
In Chrome (v70), the edit menu does not allow you to copy/paste with the Terminal. In fact, the entire pull down menu is greyed out.

**Comment by Tyriar** [comment]
@fonnesbeck I checked the xterm.js demo and it seems to work fine, so this must be a problem with jupyter (maybe CSS changes are missing?)

**Comment by jasongrout** [comment]
Thinking a bit more about William's suggestion above:

I think having ctrl-c depend on whether text is selected or not is brilliant. +1.

There *is* a terminal interpretation for ctrl-v, which is to treat the next character literally (e.g., pressing `ctrl-v ctrl-h` inserts a literal ctrl-h in the text, instead of actually interpreting it and doing a backspace). I only use this occasionally, usually when searching for `^M` line endings in vim, and  most people probably never use it. However, there may be some people who complain if ctrl-v suddenly pastes instead of doing the standard terminal thing.

**Comment by tgeorgeux** [comment]
I don't think it hurts to go forward with the contextual use of ctrl-v as discussed above. If it's causing a problem we can either roll back the change, or offer a setting to disable it.

It's hard to tell which use case is more common without any data on the topic, but I think merging and iterating is acceptable in this case. Thoughts @jasongrout ?

**Comment by jasongrout** [comment]
@ivanov also points out that, in addition to inserting characters, ctrl-v is used in vi to enter block select (http://vimdoc.sourceforge.net/htmldoc/visual.html). He said he uses the context menu or application menu to paste outside text into a terminal.

Also, ctrl-v is used for scrolling in emacs, and used in other command sequences as well. There is a cua-mode for emacs that changes ctrl-v to paste, though.

So I think if we do make ctrl-v paste, we should have a setting to disable it.

**Comment by oudarjya718** [comment]
ctrl + shift + v isnt also working

**Comment by saulshanabrook** [comment]
TLDR: Different copy and paste shortcuts for the terminal on Windows, Mac OS, Linux. Someone on here did a survey of the different versions. 

We could change how we do copy paste in the terminal or we could document what is there to make it clear how it works on the different operating systems.

The first step would be to verify on each platform what the copy paste shortcut is. This has to do with the server platform possibly not the client. Then we can decide whether to change these or just document what we have.

William's workaround is that if we have selected text and you hit CTRL C in unix, then you intercept it.

1. Document what we have for 1.0.
2. Implement William Stein's workaround for 1.1
